`PerspectiveWidget` HTML export support in Jupyter

tomjakubowski opened the pull request.

tomjakubowski commented on python/perspective/perspective/__init__.py:
```diff
@@ -16,4 +16,8 @@
 from .handlers import *
 from .nbextension import _jupyter_nbextension_paths
-from .widget import *
+from .widget.widget import PerspectiveWidget
```
this is an affordance to make `%autoreload` magic work better. happy to revert

timkpaine replied:
We need to go through and make sure all our * imports pull from files with `__all__`, this is a good first issue I'll put in another ticket

timkpaine replied:
https://github.com/finos/perspective/issues/2419

tomjakubowski replied:
reverted this

timkpaine commented on python/perspective/perspective/templates/bruce.html.jinja:
```diff
@@ -0,0 +1,40 @@
+<script type="module" src="https://cdn.jsdelivr.net/npm/@finos/perspective/dist/cdn/perspective.js"></script>
```
Should we try to parameterize these to the same versions as used by the client?

tomjakubowski replied:
yeah for sure

timkpaine commented on python/perspective/perspective/widget/widget.py:
```diff
@@ -514,4 +535,31 @@ def _make_load_message(self):
+        b64_data = base64.encodebytes(data)
+
+        jinja_env = jinja2.Environment(
+            loader=jinja2.PackageLoader("perspective"),
```
Inject versions to load from jsdelivr here

timkpaine commented on python/perspective/perspective/widget/widget.py:
```diff
@@ -514,4 +535,31 @@ def _make_load_message(self):
+            loader=jinja2.PackageLoader("perspective"),
+            autoescape=jinja2.select_autoescape(),
+        )
+        template = jinja_env.get_template("bruce.html.jinja")
```
Can inject version here

timkpaine commented on python/perspective/perspective/widget/widget.py:
```diff
@@ -514,4 +535,31 @@ def _make_load_message(self):
+                viewer_id=viewer_id,
+                viewer_attrs=viewer_attrs,
+                b64_data=b64_data.decode("utf-8"),
+            )
```
Version=whatever so we pull same js assets

timkpaine commented on python/perspective/setup.py:
```diff
@@ -66,7 +66,7 @@
 requires_aiohttp = ["aiohttp>=3,<4"]
 
-requires_jupyter = ["jupyterlab>=3.2,<4"]
+requires_jupyter = ["jupyterlab>=3.2,<4", "Jinja2>=3.0,<4"]
```
Use 2 as lower bound if you can

tomjakubowski commented:
As it is currently, rendering the HTML output is unconditional and this will add some bloat to all .ipynb using `PerspectiveWidget`. Every render of a PerspectiveJupyterWidget adds the HTML output to its cell's output in the `.ipynb` . This includes a base64-serialized arrow of every viewer's data, without deduplication, and a few lines of Javascript and HTML to glue it together.

We can make this configurable and/or optional in Perspective, for users who don't care to support exporting. Would a global setting or per-widget setting make sense?

timkpaine commented:
per-widget, what about if I stream data in and call update() a bunch of times, will it reflect the latest?

tomjakubowski commented:
I don't have very much experience doing streaming perspectives but I'll check. I think it reflects whatever was in the table at the time the cell first rendered, which is probably undesirable.

tomjakubowski commented:
Added a class variable on `PerspectiveWidget` to gate this new feature behind an undocumented flag while we're tooling on it. Also fixed a bug in datagrid

It can be set on the class to make all widgets render HTML output, or on an instance to make only that widget do it

texodus commented on python/perspective/perspective/widget/widget.py:
```diff
@@ -201,6 +203,8 @@ class PerspectiveWidget(DOMWidget, PerspectiveViewer):
     _view_module = Unicode("@finos/perspective-jupyterlab").tag(sync=True)
     _view_module_version = Unicode("~{}".format(__version__)).tag(sync=True)
 
+    _enable_experimental_html_export = False
```
It's ok to make this non-default but let's make it a real API. My proposal is two-fold:
1. Environment variable `PSP_JUPYTER_HTML_EXPORT` whose presence enables this behavior, so `nbconvert` + `nbclient` can be called from the command line.
2. A `import set_jupyter_html_export from perspectve` method which sets this env variable for the current process, so the behavior can be defined in the notebook itself (alternatively they could just set the env var but I think this is cleaner and parallels e.g. `set_num_threads()` global config APIs we already expose).

tomjakubowski replied:
Left is running jupyterlab with `PSP_JUPYTER_HTML_EXPORT=1`, right is running without the environment variable set. It would have been cool including a cell with the value of `os.environ.get("PSP_JUPYTER_HTML_EXPORT", None)` but I didn't think of it in time.

<img width="1452" alt="Screenshot 2023-11-14 at 3 49 59 PM" src="https://github.com/finos/perspective/assets/667901/01a7f462-c9a4-4181-aee0-9bfa3083a790">

texodus commented on python/perspective/perspective/templates/exported_widget.html.jinja:
```diff
@@ -0,0 +1,41 @@
+        const viewerAttrs = {{ viewer_attrs | tojson }};
+
+        // Create a new worker, then a new table promise on that worker.
+        const table = await perspective.worker().table(data.buffer);
```
Use [`shared_worker`](https://github.com/finos/perspective/blob/master/packages/perspective/src/js/perspective.browser.js#L195) or you will get a WebWorker per cell.

TODO we should probably add documentation for `shared_worker` and friends as we apparently don't yet generate JsDocs anymore for this module.

texodus commented on python/perspective/setup.py:
```diff
@@ -66,7 +66,7 @@
 requires_aiohttp = ["aiohttp>=3,<4"]
 
-requires_jupyter = ["jupyterlab>=3.2,<4"]
+requires_jupyter = ["jupyterlab>=3.2,<4", "Jinja2>=2.0,<4"]
```
IMO this is just a regular requirement. `requires_jupyterlab` is used to install Jupyter Lab in CI, real users will likely already have this installed in their environment and would never think to install the undocumented `pip install perspective-python[jupyterlab]`, making this extra `Jinja` dep easy to miss.

Also please pin this in the CI `requirements` files.

tomjakubowski replied:
I misunderstood what requires_jupyter was for… appreciate the explanation!

tomjakubowski replied:
oops, need to pin this in the requirements file.

tomjakubowski replied:
I bumped the deps using the new update_python script

It made a really noisy diff which was made a lot cleaner once I sorted the lines in the updated requirements-*.txt files. Maybe this is something we could add to the update_python script? (or to some documentation about how to update deps, which I'm happy to write)

tomjakubowski replied:
Added sorting to the requirements.txt updater

texodus reviewed (changes requested):
Thanks for the PR!
